Dockerfile

```diff
@@ -1,29 +1,115 @@
+    postgresql-dev \
+    sqlite-dev \
+    tzdata \
+    yarn
+
+# Create a generic Gemfile with only Rails.
+RUN echo -e "source 'https://rubygems.org'\ngem 'rails'" > /Gemfile
+# Install Rails, quietly, in 4 threads.
+RUN bundle install --quiet -j4
+
+# Uses the image above as the base for a builder image.
+# This will **exclude** development and testing gems.
+FROM rails as rails-builder
+
+# Set the value of `pwd` for all following operations.
+WORKDIR /app/
+
+# Add the Gemfile and Gemfile.lock from this application.
+COPY Gemfile Gemfile.lock /app/
+
+# Installs gems, quietly, in 4 threads, and retry if broken.
+RUN bundle install --quiet -j4 --retry 3 \
+    # Remove unneeded files (cached *.gem, *.o, *.c)
+    && rm -rf /usr/local/bundle/cache/*.gem \
+    && find /usr/local/bundle/gems/ -name "*.c" -delete \
+    && find /usr/local/bundle/gems/ -name "*.o" -delete
+
+# Expose port 5000 from the final image.
+# This is important if you deploy to Gitlab, otherwise it does
+# not matter which port we use here.
+EXPOSE 5000
+
+# When starting this container, run the rails server
+# binding to all IP's, and port 5000.
+CMD bundle exec rails server -b 0.0.0.0 -p 5000
+
+# This image is used to build assets and delete side effects.
+# This will **include** development and test gems.
+FROM rails-builder AS builder
+
+# Copy in the Gemfile and Gemfile.lock files in.
+COPY Gemfile Gemfile.lock /app/
+# Delete the existing bundle.
+RUN rm -rf /usr/local/bundle \
+    # Configure bundle to freeze gems.
+    && bundle config --global frozen 1 \
+    # Configure bundle to exclude development and test environments.
+    && bundle config set without 'development test' \
+    # Installs gems, quietly, in 4 threads, and retry if broken.
+    && bundle install --quiet -j4 --retry 3 \
+    # Remove unneeded files (cached *.gem, *.o, *.c)
+    && rm -rf /usr/local/bundle/cache/*.gem \
+    && find /usr/local/bundle/gems/ -name "*.c" -delete \
+    && find /usr/local/bundle/gems/ -name "*.o" -delete
 
-# Bundling
-COPY Gemfile* ./
-RUN gem install bundler:2.1.4 && bundle install
+# Add the npm package.json and yarn.lock to install packages.
+COPY package.json yarn.lock /app/
+# Use yarn to install node dependencies.
+RUN yarn install
 
-# Building /node_modules
-COPY yarn.lock .
-COPY package.json .
-RUN yarn install --check-files
+# Copy in everything from the current directory.
+COPY ./ /app/
+
+# Compile assets in Production mode.
+RUN RAILS_ENV=production SECRET_KEY_BASE=not_for_prod bundle exec rake assets:precompile
+
+# Clean up temporary files.
+RUN rm -rf node_modules tmp/cache vendor/assets lib/assets spec
+
+# Final, production, image.  Minimum footprint.
+FROM ruby:2.7-alpine AS final
+
+# Set rails to production mode.
+ENV RAILS_ENV production
+# Tell Rails to log to STDOUT so it shows in Docker logs.
+ENV RAILS_LOG_TO_STDOUT true
+# Tell Rails to server static files, rather than using a
+# reverse proxy. 
+ENV RAILS_SERVE_STATIC_FILES true
```

Comment:
What does this mean?

Reply:
By default Rails will not serve static files from the `public` directory in Production.  Turning this on removes the need for a reverse proxy in front of the Rails server.  From the [Rails docs](https://guides.rubyonrails.org/configuring.html):

> config.public_file_server.enabled configures Rails to serve static files from the public directory. This option defaults to true, but in the production environment it is set to false because the server software (e.g. NGINX or Apache) used to run the application should serve static files instead. If you are running or testing your app in production using WEBrick (it is not recommended to use WEBrick in production) set the option to true. Otherwise, you won't be able to use page caching and request for files that exist under the public directory.

Reply:
Okay! Thanks for the context. I'll look at this again when we move the production build on Heroku to use a Docker image.